VirtualMachinePool Design Proposal

davidvossel opened the pull request:
This is a design proposal for a VirtualMachinePool abstraction to manage replicas of stateful VirtualMachines at scale.

davidvossel commented:
/cc @rthallisey

kwiesmueller commented on design-proposals/vm-pool.md:
```diff
@@ -0,0 +1,194 @@
+
+* **VirtualMachineRef** - (Required) A structure that points to a VMConfig object
+* **Replicas** - (Required) An integer representing the desired number of VM replicas
+* **MaxUnavailable**  - (Optional) (Defaults to 25%) Integer or string pointer, that when set represents either a percentage or number of VMs in a pool that can be unavailable (ready condition false) at a time during automated update.
```
Here we could borrow from https://kubernetes.io/docs/reference/kubernetes-api/workload-resources/stateful-set-v1/#StatefulSetSpec to be consistent and keep our options open.
So move anything regarding updates into a `updateStrategy` field as well instead of having it top-level.
That also matches the fact that the design already contains a separate `disruptionStrategy`.

davidvossel replied:
the disruption strategy was meant to combine how to handle both updates and scale-in. Are you saying you think those should be treated separately?

kwiesmueller commented on design-proposals/vm-pool.md:
```diff
@@ -0,0 +1,194 @@
+	* **Opportunistic** - Opportunistic scale-in and update of VMs which are in a halted state.
+		* **ScaleInStatePreservation** - (Optional) specifies if and how to preserve state of VMs selected for scale-in.
+			* **Disabled** - (Default) all state for VMs selected for scale-in will be deleted
+			* **Offline** - PVCs for VMs selected for scale-in will be preserved and reused on scale-out (decreases provisioning time during scale out)
```
Would that mean name generation has to be consistent so the generated names for connected resources (storage) match a scaled-in VM when it happens?
From what I see above name generation would be incremental so it should be fine, just want to make sure.

kwiesmueller commented on design-proposals/vm-pool.md:
```diff
@@ -0,0 +1,194 @@
+* **VirtualMachine [VM]** - Refers specifically to the KubeVirt VirtualMachine api object. 
+* **VirtualMachinePool [VMPool]** - KubeVirt API for scaling out/in replicas of KubeVirt VirtualMachines
+* **VirtualMachineConfig [VMConfig]** - KubeVirt API representing a VirtualMachine spec which is referenced by a VirtualMachinePool.
+* **Scale-out and Scale-in** - Terms to describe the action of adding/removing to the count of VMs within a VMPool.
```
Are Scale-in and Scale-out common terms for that kind of pattern?
At first reading I mixed them up as I read scale-in as "scaling VMs into the pool -> growing" which is obviously wrong.
What about Scale-Up and down, I think that's also very known terminology.

davidvossel replied:
scale-out and scale-up  have been used interchangeably in the past. I see that scale-out is becoming more consistent in conversations where we're scaling horizontally.  scale-up can also refer to vertical scaling, which is why i'm beginning to prefer scale-out

kwiesmueller commented on design-proposals/vm-pool.md:
```diff
@@ -0,0 +1,194 @@
+	* **Proactive** - Proactive scale-in and update by forcing VMs to shutdown during scale-in and restart during update.
+		* **ScaleInPriority** - The priority in which VM instances are selected for proactive scale-in
+			* **Random** (Default) - Pick VMs by random for scale-in first
+			* **Oldest** - Pick oldest VMs by creation time for scale-in first
```
How about newest as well?

davidvossel replied:
sure, that could make sense, i picked two here just to show that it's an enum

kwiesmueller commented on design-proposals/vm-pool.md:
```diff
@@ -0,0 +1,194 @@
+
+## Manually Detaching VM from VMPool
+
+A VM in a VMPool can be detached from a VMPool by removing the owner reference. This removes that VM from being actively managed by the VMPool.
```
These operations could deserve convenience subresources on the Pool API.
And for the future it could also be desirable to instead of just detaching which is kind of like claiming, actually removing a VM from the pool by (live-)migrating it into a cloned VM with a new name.
That would let the Pool refill the name if it is certain a VM never goes back into the pool.

davidvossel replied:
> These operations could deserve convenience subresources on the Pool API.

that could make sense

> And for the future it could also be desirable to instead of just detaching which is kind of like claiming, actually removing a VM from the pool by (live-)migrating it into a cloned VM with a new name.
That would let the Pool refill the name if it is certain a VM never goes back into the pool.

yup, there wouldn't be anything preventing someone from cloning a VM managed by a VM pool.

kwiesmueller commented on design-proposals/vm-pool.md:
```diff
@@ -0,0 +1,194 @@
+
+When managing VMs at large scale, it is useful to automate the recovery of VMs which continually fail to reach a ready state. This automatically fixes scenarios where a VM’s state has somehow been corrupted and needs to be completely refreshed.
+
+By setting `AutohealingStrategy=ReprovisionOnFailure` on the VMPool’s spec, VMs which reach an unexpected failure state will automatically be completely deleted (including persistent storage) and re-provisioned. This allows for auto recovery in application scenarios that can withstand such an action.
```
We should probably also have some sort of CrashLoopBackoff just to make sure we don't contribute to any strain that might cause provisioning to fail in the first place.

davidvossel replied:
yup, i'll note that.

kwiesmueller reviewed (commented):
Thanks for the great write-up. Everything I remember discussing is in here.

rthallisey commented on design-proposals/vm-pool.md:
```diff
@@ -0,0 +1,194 @@
+Each VM’s PVCs are preserved for future scale out
+	* **Proactive** - Proactive scale-in and update by forcing VMs to shutdown during scale-in and restart during update.
+		* **ScaleInPriority** - The priority in which VM instances are selected for proactive scale-in
+			* **Random** (Default) - Pick VMs by random for scale-in first
```
Can you have both Proactive and ScaleInStatePreservation?  In other words, after all the stopped VMIs are removed, can the user control how the next set of VMIs are selected for removal?

Maybe `DisruptionStrategy` can be an ordered list?

davidvossel replied:
> Can you have both Proactive and ScaleInStatePreservation? 

yes, this would result in VMs which are removed during scale-in orphaning their PVCs, and then those VMs which are re-created during scale-out reclaim those orphaned PVCs.

> in other words, after all the stopped VMIs are removed, can the user control how the next set of VMIs are selected for removal?

The disruptionStrategy can be mutated at any time on the VMPool, which then impacts how future scale-in and update events occur. Is that what you're looking for?

rthallisey replied:
I had a slightly different take on policy.  I was thinking we could define one "BasePolicy", which is a policy that's always true (e.g. oldest or newest), then have a bunch of tuning policies.  "BasePolicy" would be what you have now for `Priority`.

BasePolicy - prioritize by Name or Age.  A base policy is always true so only 1 can be chosen at at time
- lowest (vm-1, vm-2, ect...)
- highest (vm-n, vm-(n-1), ect...)
- oldest
- newest
- random

TuningPolicies:
LabelPolicy - prioritize by label.  Accepts a list of labels
- Scale-in order follows the ordered list of labels. Allows admins/another controller to dynamically label VMIs to control ordering. This can be a "plug-in" for admins to inform the pool about guest information

NodePolicy - prioritize by node. Accepts a list of nodes
- Scale-in by node

Example: 
```yaml
- Priority:
    - LabelPolicy :
         - unimportant-vmis
         - not-very-important-vmis
         - important
         - super-important
     - NodePolicy:
         - node01
         - node02
     - BasePolicy: "oldest"
```
The scale-in Priority order would be:
1. VMIs labeled `unimportant-vmis`
2. VMIs labeled `not-very-important-vmis`
3. VMIs labeled `important`
4. VMIs labeled `super-important`
5. VMIs on node01
6. VMIs on node02
7. Oldest VMIs

> Maybe DisruptionStrategy can be an ordered list?
>
This is what I meant by using an ordered list.  There are a few powerful tunings we can take advance of to provide the fine-grained scale-in control.  WDYT?

davidvossel replied:
> This is what I meant by using an ordered list. There are a few powerful tunings we can take advance of to provide the fine-grained scale-in control. WDYT?

yeah, makes sense to me. Do we really need a `basePolicy` here though. For example wouldn't something like what i have below work, where there's no basePolicy but oldest is just another policy.

```
- Priority:
    - labelPolicy:
         - unimportant-vmis
         - not-very-important-vmis
         - important
         - super-important
     - nodePolicy:
         - node01
         - node02
     - OldestPolicy: {}
```

I guess the weird thing here would be if someone attempted to use the OldestPolicy as the first item in the list. That would make the rest of the items never get hit. A validation webhook could catch that though i suppose.

rthallisey replied:
Ya that would work.  The name BasePolicy is so we have a name for a group of these policies that will always match.  Maybe it needs a different name?  I just thought it was easier to communicate the point.

davidvossel replied:
> The name BasePolicy is so we have a name for a group of these policies that will always match. Maybe it needs a different name? I just thought it was easier to communicate the point.

ah that makes sense. I'll take some time to think about the naming and structure then revise the doc to take all this into account.

great feedback!

davidvossel replied:
@rthallisey I updated the doc. I called the struct `SelectionPolicy`. Within that struct, people have the option to set both a list of orderedPolicies and/or a basePolicy. 

when set, the ordered policies are matched first, in order, then if a catch all "basePolicy" is set, that will act as a catch all. If no basePolicy is set, no vms will get matched and scale-in will block until an item in the orderedPolicy matches.

Below is an example. The priority for scale-in is, by label first, then by node, then base "oldest". The priority for update in this example has no ordered list and  is simply by "oldest" 

```
apiVersion: kubevirt.io/v1
kind: VirtualMachinePool
metadata:
  name: my-vm-pool
spec:
  VirtualMachineConfigRef:
    name: my-template
  replicas: 100
  maxUnavailable: 10
  scaleInStrategy:
    proactive:
      selectionPolicy:
        orderedPolicies:
          - labelSelection
            - non-important-vms
          - nodeSelection
            - node2
            - node3
        basePolicy: "Oldest"
      statePreservation: Offline
  updateStrategy:
    proactive:
      selectionPolicy:
        basePolicy: "Oldest"
```

rthallisey commented on design-proposals/vm-pool.md:
```diff
@@ -0,0 +1,194 @@
+A VMIReplicaSet makes sense for a user who wants to manage stateless VMI objects using similar patterns to how Kubernetes ReplicaSets manage pods.
+
+A VMPool makes sense for a user who wants to manage stateful VM objects using similar patterns found in Iaas public clouds, like AWS, GCP, Azure. In this way, a VMPool is more similar to an AWS AutoscalingGroup or a GCP ManagedInstanceGroup than it is to any core Kubernetes API.
+
```
Can we define some metrics?
- The number of VMIs restarted in a Pool
- The number of VMIs detached from a Pool
- Pool VMI perf metrics (A Pool can add a label to the phaseTransitionTimes so we can associate per pool)

VirrageS commented on design-proposals/vm-pool.md:
```diff
@@ -0,0 +1,194 @@
+
+## VirtualMachineConfig (VMConfig) API
+
+The VMConfig api contains a VirtualMachineSpec structure that is identical to a VM object, which makes it similar to a VM object itself. However there are some key differences here between the two objects.
```
```suggestion
The VMConfig API contains a VirtualMachineSpec structure that is identical to a VM object, which makes it similar to a VM object itself. However there are some key differences here between the two objects.
```

davidvossel replied:
+1

VirrageS commented on design-proposals/vm-pool.md:
```diff
@@ -0,0 +1,194 @@
+
+## VirtualMachinePool (VMPool) API
+
+The VMPool api represents all the tunings necessary for managing a pool of stateful VMs. This API is very simple and leverages the VMConfig object for most of the heavy lifting pertaining to what the VM instances within the pool actually look like.
```
```suggestion
The VMPool API represents all the tunings necessary for managing a pool of stateful VMs. This API is very simple and leverages the VMConfig object for most of the heavy lifting pertaining to what the VM instances within the pool actually look like.
```

davidvossel replied:
+1

VirrageS commented on design-proposals/vm-pool.md:
```diff
@@ -0,0 +1,194 @@
+* **MaxUnavailable**  - (Optional) (Defaults to 25%) Integer or string pointer, that when set represents either a percentage or number of VMs in a pool that can be unavailable (ready condition false) at a time during automated update.
+* **NameGeneration** - (Optional) Specifies how objects within a pool have their names generated
+	* **VMPrefix (Defaults to VMPool.Name)** - String representing the prefix to use for when naming VMs.
+	* **AppendPostfixToSecretReferences: (default false)** - Boolean that indicates if VM’s unique postfix should be appended to references to Secrets in the VMI’s Volumes list. This is useful when needing to pre-generate unique secrets for VMs within a pool.
```
```suggestion
	* **AppendPostfixToSecretReferences (default false)** - Boolean that indicates if VM’s unique postfix should be appended to references to Secrets in the VMI’s Volumes list. This is useful when needing to pre-generate unique secrets for VMs within a pool.
```

VirrageS commented on design-proposals/vm-pool.md:
```diff
@@ -0,0 +1,194 @@
+
+The VMPool api represents all the tunings necessary for managing a pool of stateful VMs. This API is very simple and leverages the VMConfig object for most of the heavy lifting pertaining to what the VM instances within the pool actually look like.
+
+The VMPools spec contains the following tunings and values
```
There is mix between `**Something (Default/Optional)** -`, `**Something** (Default/Optional) -` and `**Something** - (Default/Optional)` in this section.

davidvossel replied:
i'll scan through this and make sure I pick a consistent convention.

VirrageS commented on design-proposals/vm-pool.md:
```diff
@@ -0,0 +1,194 @@
+metadata:                                                            
+  name: my-vm-pool                                                   
+spec:                                                                
+  VirtualMachineConfigRef:                                                                
```
```suggestion
  virtualMachineRef:
```
Also is the `kind` missing here? In the previous structure it is present but here not. Or here we assume that it will be set to default?

davidvossel replied:
i made this more consistent.

The assumption here is that the VirtualMachineConfigRef will be used, and just have a `name` field for now. This gives us the ability to use other API types for assigning VM configuration like data to a VMPool in the future. but for now VMCOnfigs are the only thing supported.

acardace commented on design-proposals/vm-pool.md:
```diff
@@ -0,0 +1,194 @@
+
+# Design
+
+The VMPool design introduces two new APIs represented as CRDs, the VirtualMachinePool (VMPool) object and the VirtualMachineConfig (VMConfig) object. For the most part, a VMConfig object is simply a VM object without a status section that is used by a VMPool to stamp out multiple replicas.
```
Why did you choose to have a separate VMConfig object instead of embedding the VMSpec in the VMPool resource? It's usually pretty handy to have everything in one place for inspecting and replicating a manifest with similar configs.

VirrageS replied:
This is explained in some of the next sections.

kwiesmueller replied:
I like that change as it allows to easily create different pools from a single base config.

davidvossel replied:
> Why did you choose to have a separate VMConfig object instead of embedding the VMSpec in the VMPool resource? It's usually pretty handy to have everything in one place for inspecting and replicating a manifest with similar configs.

a couple primary reasons off the top of my head.

- it gives VMConfigs a 1 to many relationship to VMPools. Meaning one VMConfig can be assigned to multiple VMPools. 
- it allows people to plan and stage VMPool updates. For example, someone can create different VMConfig versions. Roll out the new VMCOnfig in some dev/staging environments, then roll it out to Prod. Rolling back is as simple are assigning the previous VMConfig to the VMPool.

davidvossel replied:
@acardace i'm actually leaning towards your suggestion now and revised the document to embed the VMSpec into the VMPool.

davidvossel commented on design-proposals/vm-pool.md:
```diff
@@ -0,0 +1,194 @@
+
+## Terms
+
+* **VirtualMachine [VM]** - Refers specifically to the KubeVirt VirtualMachine api object. 
```
the pool only acts on VM objects.  VMIs are ephemeral and can't survive shutdown/restart events. 

for someone who just wants to work with VMIs, the VMIReplicaSet controller is available for them. I wrote a sort section in this doc describing the differences between VMPool and VMIReplicaSet

davidvossel commented on design-proposals/vm-pool.md:
```diff
@@ -0,0 +1,194 @@
+* Automated rollout of spec changes across and other updates across a pool of stateful VMs
+* Automated and manual scale-out and scale-in VMs associated with a pool
+* Autohealing (delete and replace) of VMs in a pool not passing health checks
+* Ability to specify unique secrets and configMap data per VM within a VMPool
```
yup, i'm adding it.

davidvossel commented on design-proposals/vm-pool.md:
```diff
@@ -0,0 +1,194 @@
+* Autohealing (delete and replace) of VMs in a pool not passing health checks
+* Ability to specify unique secrets and configMap data per VM within a VMPool
+
+## Non-Goals
```
can you provide some more details about what you mean by `controlled scale-out`?

rthallisey replied:
What I mean by controlled scale-out is having a policy that defines how scale-out happens e.g. some form of ordering.  You defined controlled scale-in in `DisruptionStrategy`, so this would be the inverse.

davidvossel commented on design-proposals/vm-pool.md:
```diff
@@ -0,0 +1,194 @@
+
+By default, VM names are generated from the VMPool’s name by appending the VMPool’s name with a sequential unique postfix. This is similar to how pods are generated from a StatefulSet’s name.
+
+## State Preservation during Scale-in
```
The naming will always be `prefix` + `sequence number`

So lets say a VMPool has 3 VMs.

- my-vm-1
- my-vm-2
- my-vm-3

and during scale-in (replicas:2) , `my-vm-2` is removed, which results in the set looking like this.

- my-vm-1
- my-vm-3

On the next scale-out event, the VMPool is going to search sequentially to fill any gaps before appending new sequence numbers, so if we expand replicas to (replicas:4), the new set will fill in the missing my-vm-2 and create a new my-vm-4

- my-vm-1
- my-vm-2 (recreated from previous state if `ScaleInPreserveState=offline`
- my-vm-3
- my-vm-4 (newly provisioned)

rthallisey replied:
I think an example exactly like what you provided here would be helpful in the doc.  It explains the scale-out naming really well

davidvossel replied:
good idea. I added this example to the doc

davidvossel commented on design-proposals/vm-pool.md:
```diff
@@ -0,0 +1,194 @@
+
+The VMPool controller should establish some default upper limits when it comes to how many VMs can be batch created and updated at a time. This is similar to the VMIReplicaSet controller's internal `BurstReplicas` global variable. By default VMIReplicaSets will only create at most 250 VMIs at a time, but this is configurable as a global setting. The VMPool controller should adopt the same pattern ...
+
+This setting is primarily meant as a way of preventing the VMPool controller from creating an unintentionally cluster DoS.
```
> Agreed. We also have APF as a backstop in k8s 1.21 onward.

that helps too. This is probably an area where we will evolve as our understanding of VMPool performance improves.

davidvossel commented on design-proposals/vm-pool.md:
```diff
@@ -0,0 +1,194 @@
+
+## Terms
+
+* **VirtualMachine [VM]** - Refers specifically to the KubeVirt VirtualMachine api object. 
```
yup, i'll revise to use consistent `API` upper casing over `api`.

davidvossel commented on design-proposals/vm-pool.md:
```diff
@@ -0,0 +1,194 @@
+## Terms
+
+* **VirtualMachine [VM]** - Refers specifically to the KubeVirt VirtualMachine api object. 
+* **VirtualMachinePool [VMPool]** - KubeVirt API for scaling out/in replicas of KubeVirt VirtualMachines
```
+1

davidvossel commented on design-proposals/vm-pool.md:
```diff
@@ -0,0 +1,194 @@
+
+## Non-Goals
+
+* Not designing a pool abstraction capable of managing multiple VM shapes within a single pool. A VMPool consists only of VMs replicated from a single config.
```
yup, making that change.

davidvossel commented on design-proposals/vm-pool.md:
```diff
@@ -0,0 +1,194 @@
+* **VirtualMachineConfig [VMConfig]** - KubeVirt API representing a VirtualMachine spec which is referenced by a VirtualMachinePool.
+* **Scale-out and Scale-in** - Terms to describe the action of adding/removing to the count of VMs within a VMPool.
+* **Detach VM** - The process of manually separating a VirtualMachine from a VirtualMachinePool.
+* **Disruption Strategy** - The policy used to define how a VMPool both selects VMs during scale-in and update.
```
i revised the disruption strategy into two new fields, updateStategy and scaleInStrategy. So the wording here has changed a bit.

davidvossel commented:
updated.

Besides addressing feedback, the big change here is i split the `disruptionStrategy` field into two fields, `scaleInStrategy` and `updateStrategy`. This allows users to control how updates are rolled out independently of how scale-in operations are performed.

rthallisey commented on design-proposals/vm-pool.md:
```diff
@@ -0,0 +1,242 @@
+
+## VirtualMachineConfig (VMConfig) API
+
+The VMConfig API contains a VirtualMachineSpec structure that is identical to a VM object, which makes it similar to a VM object itself. However there are some key differences here between the two objects.
```
Asked in SIG-scale:  Why do we want to use a VM object in the Pool and not VMIs?  In other words, a Pool is an abstraction for many objects, why can't those be VMIs instead of VMs? 

My take away was that VMs (the kubevirt API) hold state.  A VMPool is an abstraction for stateful virtual machines or those that need advanced management.  VMI replicaset would be the abstraction for stateless virtual machines.  Do you want to add a comment about this in case it comes up again?

rthallisey commented on design-proposals/vm-pool.md:
```diff
@@ -0,0 +1,242 @@
+	* **Opportunistic** - Opportunistic update of VMs which are in a halted state.
+	* **Proactive** - (Default) Proactive update by forcing VMs to restart during update.
+		* **SelectionPolicy** - (Optional) (Defaults to "random" base policy when no SelectionPolicy is configured) The priority in which VM instances are selected for proactive scale-in
+			* **OrderedPolicies** - (Optional) Ordered list of selection policies. Policies include [LabelSelector|NodeSelector]
```
From SIG-scale: During Update and scale-in, there could be VMIs that are not `Running` yet.  There could be an additional policy here to target VMIs that are not `Running` first, before deleting VMIs that are (and vice-versa) - it would improve upgrade speed.

davidvossel commented:
@acardace @kwiesmueller @VirrageS @rmohr @rthallisey 

I made a kind of fundamental change to the document and removed the VMConfig API. The VM configuration is now embedded into the VMPool using a template similar to how Deployments/StatefulSets work.

Here's my thought process. Originally I wanted someone to be able to take one VMConfig and allow it to be mapped to multiple VMPools, however I don't think my use case for that functionality is strong enough. I thought it might be useful to have a VMConfig mapped to a staging environment, then allow that exact same VMConfig to be mapped to Prod later on, but realistically that's never going to happen. Staging and Prod will never be in the same namespace (VMConfigs are namespace scoped) and these environments probably won't even be in the same cluster. My other reasoning was that it allowed people to performing their own versioning of VMConfigs separate from the VMPool which lets them easily roll back. However in reality people will likely just use a gitops flow and rollback their changes from some repo... and if they aren't doing things that way, we should probably align with the kubectl rollback functionality. 

By embedding the VM into the VMPool, we align with the k8s way of layering objects within objects (See Deployments/StatefulSets) and with that we gain the ability to leverage some k8s functionality here. For example, kubectl history/rollback would be something we could support if we wanted.
